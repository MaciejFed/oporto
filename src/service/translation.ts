import { getRandomElement } from '../common/common';
import { Adjective, Other, readAll, Verb } from '../repository/exercisesRepository';
import { Noun, Sentence } from '../repository/exercisesRepository';

export const getRandomNoun: () => Noun = () => {
  return getRandomElement(readAll().nouns);
};

export const getRandomOther: () => Other = () => {
  return getRandomElement(readAll().others);
};

export const getRandomAdjective: () => Adjective = () => {
  return getRandomElement(readAll().adjectives);
};

export const getRandomVerb: () => Verb = () => {
  return getRandomElement(readAll().verbs);
};

export const getRandomSentence: () => Sentence = () => {
  return getRandomElement(readAll().sentences);
};
