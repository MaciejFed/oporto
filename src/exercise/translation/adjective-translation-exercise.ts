import { BaseWordType, ExerciseType } from '../exercise';
import { Comparable } from '../../common/common';
import { Adjective } from '../../repository/exercises-repository';
import { getRandomAdjective } from '../../service/translation';
import { TranslationExercise, TranslationType } from './translation-exercise';

export class AdjectiveTranslationExercise extends TranslationExercise implements Comparable {
  exerciseType: ExerciseType;
  adjective: Adjective;
  gender: 'masculine' | 'feminine';
  number: 'singular' | 'plural';

  constructor() {
    super();
    this.exerciseType = 'AdjectiveTranslation';
    this.adjective = getRandomAdjective();
    this.gender = Math.random() > 0.5 ? 'masculine' : 'feminine';
    this.number = Math.random() > 0.5 ? 'singular' : 'plural';
  }

  static new(
    adjective: Adjective,
    translationType: TranslationType,
    gender: 'masculine' | 'feminine',
    number: 'singular' | 'plural'
  ): AdjectiveTranslationExercise {
    const adjectiveTranslationExercise = new AdjectiveTranslationExercise();
    adjectiveTranslationExercise.adjective = adjective;
    adjectiveTranslationExercise.translationType = translationType;
    adjectiveTranslationExercise.gender = gender;
    adjectiveTranslationExercise.number = number;

    return adjectiveTranslationExercise;
  }

  isTranslationSubjectEqual(translationExercise: TranslationExercise): boolean {
    if (!(translationExercise instanceof AdjectiveTranslationExercise)) return false;
    return this.adjective.masculine.singular === translationExercise.adjective.masculine.singular;
  }

  getBodyPrefix(): string {
    return this.isTranslationToPortuguese() ? `Portuguese: ${this.getBodySuffix()}` : 'English: ';
  }

  getBodySuffix = () => ` [${this.gender}, ${this.number}]`;

  getDescription = () => {
    if (this.isTranslationToPortugueseFromHearing()) return 'Listen...';
    if (this.isTranslationToPortuguese()) {
      return `English: ${this.adjective.english}`;
    }
    return `Portuguese: ${this.adjective.masculine.singular}`;
  };

  getTranslation = () => this.adjective.english;

  getCorrectAnswer = () =>
    this.isTranslationToPortuguese() ? this.adjective[this.gender][this.number] : this.adjective.english;

  isAnswerCorrect(answer: string): boolean {
    return this.getCorrectAnswer().toLowerCase() === answer.toLowerCase();
  }

  getRetryPrompt = () =>
    this.isTranslationToPortuguese() ? this.getCorrectAnswer() : this.adjective[this.gender][this.number];

  getBaseWord() {
    return this.adjective;
  }

  getBaseWordType(): BaseWordType {
    return BaseWordType.ADJECTIVE;
  }

  getBaseWordAsString(): string {
    return this.getBaseWord().masculine.singular;
  }

  equal = (other: AdjectiveTranslationExercise) =>
    other.exerciseType === 'AdjectiveTranslation' &&
    this.gender === other.gender &&
    this.number === other.number &&
    this.adjective.masculine.singular === other.adjective.masculine.singular &&
    this.adjective.masculine.singular === other.adjective.masculine.singular &&
    this.adjective.feminine.plural === other.adjective.feminine.plural &&
    this.adjective.masculine.plural === other.adjective.masculine.plural &&
    this.translationType === other.translationType;
}
