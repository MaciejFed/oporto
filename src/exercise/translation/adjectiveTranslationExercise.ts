import { ExerciseType } from '../exercise';
import { Comparable } from '../../common/common';
import { Adjective, Verb } from '../../repository/exercisesRepository';
import { getRandomAdjective } from '../../service/translation';
import { TranslationExercise, TranslationType } from './translationExercise';

export class AdjectiveTranslationExercise extends TranslationExercise implements Comparable {
  exerciseType: ExerciseType;
  adjective: Adjective;
  gender: 'masculine' | 'feminine';
  number: 'singular' | 'plural';

  constructor() {
    super();
    this.exerciseType = 'AdjectiveTranslation';
    this.adjective = getRandomAdjective();
    this.gender = Math.random() > 0.5 ? 'masculine' : 'feminine';
    this.number = Math.random() > 0.5 ? 'singular' : 'plural';
  }

  static new(
    adjective: Adjective,
    translationType: TranslationType,
    gender: 'masculine' | 'feminine',
    number: 'singular' | 'plural'
  ): AdjectiveTranslationExercise {
    const adjectiveTranslationExercise = new AdjectiveTranslationExercise();
    adjectiveTranslationExercise.adjective = adjective;
    adjectiveTranslationExercise.translationType = translationType;

    return adjectiveTranslationExercise;
  }

  isTranslationSubjectEqual(translationExercise: TranslationExercise): boolean {
    if (!(translationExercise instanceof AdjectiveTranslationExercise)) return false;
    return this.adjective.masculine.singular === translationExercise.adjective.masculine.singular;
  }

  getExerciseBodyPrefix(): string {
    return this.isTranslationToPortuguese() ? 'Portuguese: ' : 'English: ';
  }

  getExerciseBodySuffix = () => ` [${this.gender}, ${this.number}]`;

  getExerciseDescription = () => {
    if (this.isTranslationToPortugueseFromHearing()) return 'Listen...';
    if (this.isTranslationToPortuguese()) {
      return `English: ${this.adjective.english}`;
    }
    return `Portuguese: ${this.adjective.masculine.singular}`;
  };

  getExerciseExplanation = () => (this.isTranslationToPortugueseFromHearing() ? this.adjective.english : undefined);

  getCorrectAnswer = () =>
    this.isTranslationToPortuguese() ? this.adjective[this.gender][this.number] : this.adjective.english;

  checkAnswerCorrect(answer: string): boolean {
    return this.getCorrectAnswer().toLowerCase() === answer.toLowerCase();
  }

  getRepeatAnswerPhrase = () =>
    this.isTranslationToPortuguese() ? this.getCorrectAnswer() : this.adjective[this.gender][this.number];

  equal = (other: AdjectiveTranslationExercise) =>
    other.exerciseType === 'AdjectiveTranslation' &&
    this.adjective.masculine.singular === other.adjective.masculine.singular &&
    this.translationType === other.translationType;
}
