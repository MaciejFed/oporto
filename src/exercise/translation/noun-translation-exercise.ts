import { ExerciseType } from '../exercise';
import { Comparable } from '../../common/common';
import { Noun } from '../../repository/exercises-repository';
import { getRandomNoun } from '../../service/translation';
import { TranslationExercise, TranslationType } from './translation-exercise';

export class NounTranslationExercise extends TranslationExercise implements Comparable {
  exerciseType: ExerciseType;
  noun: Noun;

  constructor() {
    super();
    this.exerciseType = 'NounTranslation';
    this.noun = getRandomNoun();
  }

  static new(noun: Noun, translationType: TranslationType): NounTranslationExercise {
    const nounExercise = new NounTranslationExercise();
    nounExercise.noun = noun;
    nounExercise.translationType = translationType;

    return nounExercise;
  }

  isTranslationSubjectEqual(translationExercise: TranslationExercise): boolean {
    if (!(translationExercise instanceof NounTranslationExercise)) return false;
    return this.noun.portuguese.word === translationExercise.noun.portuguese.word;
  }

  getExerciseBodyPrefix(): string {
    return this.isTranslationToPortuguese() ? 'Portuguese: ' : 'English: ';
  }

  getExerciseBodySuffix = () => '';

  getExerciseDescription = () => {
    if (this.isTranslationToPortugueseFromHearing()) return 'Listen...';
    if (this.isTranslationToPortuguese()) {
      return `English: ${this.noun.english}`;
    }
    return `Portuguese: ${this.getWordWithGender()}`;
  };

  getExerciseTranslation = () => (this.isTranslationToPortugueseFromHearing() ? this.noun.english : undefined);

  getCorrectAnswer = () => (this.isTranslationToPortuguese() ? this.getWordWithGender() : this.noun.english);

  checkAnswerCorrect(answer: string): boolean {
    return this.getCorrectAnswer().toLowerCase() === answer.toLowerCase();
  }

  getRepeatAnswerPhrase = () => (this.isTranslationToPortuguese() ? this.getCorrectAnswer() : this.getWordWithGender());

  equal = (other: NounTranslationExercise) =>
    other.exerciseType === 'NounTranslation' &&
    this.noun.portuguese.word === other.noun.portuguese.word &&
    this.translationType === other.translationType;

  getWordWithGender() {
    if (this.noun.portuguese.gender === 'none') return this.noun.portuguese.word;
    return `${this.noun.portuguese.gender === 'masculine' ? 'o' : 'a'} ${this.noun.portuguese.word}`;
  }
}
