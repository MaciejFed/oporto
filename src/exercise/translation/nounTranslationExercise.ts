import { Exercise, ExerciseType } from '../exercise';
import { Comparable } from '../../common/common';
import { Noun } from '../../repository/exercisesRepository';
import { getRandomNoun } from '../../service/translation';
import { TranslationExercise } from './translationExercise';

export class NounTranslationExercise extends TranslationExercise implements Comparable {
  exerciseType: ExerciseType;
  correctAnswer: string;
  noun: Noun;

  constructor() {
    super();
    this.exerciseType = 'NounTranslation';
    this.noun = getRandomNoun();
    this.correctAnswer = this.getCorrectAnswer();
  }

  isTranslationSubjectEqual(translationExercise: TranslationExercise): boolean {
    if (!(translationExercise instanceof NounTranslationExercise)) return false;
    return this.noun.portuguese.word === translationExercise.noun.portuguese.word;
  }

  getExerciseBodyPrefix(): string {
    return this.isTranslationToPortuguese() ? 'Portuguese: ' : 'English: ';
  }

  getExerciseBodySuffix = () => '';

  getExerciseDescription = () => {
    if (this.isTranslationToPortugueseFromHearing()) return 'Listen...';
    if (this.isTranslationToPortuguese()) {
      return `English: ${this.noun.english}`;
    }
    return `Portuguese: ${this.getWordWithGender()}`;
  };

  getExerciseExplanation = () => (this.isTranslationToPortugueseFromHearing() ? this.noun.english : undefined);

  getCorrectAnswer = () => (this.isTranslationToPortuguese() ? this.getWordWithGender() : this.noun.english);

  checkAnswerCorrect(answer: string): boolean {
    return this.getCorrectAnswer().toLowerCase() === answer.toLowerCase();
  }

  getRepeatAnswerPhrase = () => (this.isTranslationToPortuguese() ? this.correctAnswer : this.getWordWithGender());

  equal = (other: NounTranslationExercise) =>
    other.exerciseType === 'NounTranslation' &&
    this.noun.portuguese.word === other.noun.portuguese.word &&
    this.translationType === other.translationType;

  getWordWithGender() {
    if (this.noun.portuguese.gender === 'none') return this.noun.portuguese.word;
    return `${this.noun.portuguese.gender === 'masculine' ? 'o' : 'a'} ${this.noun.portuguese.word}`;
  }
}
