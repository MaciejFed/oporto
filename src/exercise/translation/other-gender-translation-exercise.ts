import { BaseWordType, ExerciseType } from '../exercise';
import { Comparable } from '../../common/common';
import { Adjective, OtherWithGender } from '../../repository/exercises-repository';
import { getRandomAdjective, getRandomOtherWithGender } from '../../service/translation';
import { TranslationExercise, TranslationType } from './translation-exercise';

export class OtherGenderTranslationExercise extends TranslationExercise implements Comparable {
  exerciseType: ExerciseType;
  other: OtherWithGender;
  base: boolean;
  gender: 'masculine' | 'feminine';
  number: 'singular' | 'plural';

  constructor() {
    super();
    this.exerciseType = 'OtherWithGenderTranslation';
    this.other = getRandomOtherWithGender();
    this.gender = Math.random() > 0.5 ? 'masculine' : 'feminine';
    this.number = Math.random() > 0.5 ? 'singular' : 'plural';
    this.base = false;
  }

  static new(
    other: OtherWithGender,
    translationType: TranslationType,
    gender: 'masculine' | 'feminine',
    number: 'singular' | 'plural',
    base?: boolean
  ): OtherGenderTranslationExercise {
    const otherGenderTranslationExercise = new OtherGenderTranslationExercise();
    otherGenderTranslationExercise.other = other;
    otherGenderTranslationExercise.translationType = translationType;
    otherGenderTranslationExercise.gender = gender;
    otherGenderTranslationExercise.number = number;
    otherGenderTranslationExercise.base = base || false;

    return otherGenderTranslationExercise;
  }

  isTranslationSubjectEqual(translationExercise: TranslationExercise): boolean {
    if (!(translationExercise instanceof OtherGenderTranslationExercise)) return false;
    return this.other.portuguese.singular.masculine === translationExercise.other.portuguese.singular.masculine;
  }

  getBodyPrefix(): string {
    return this.isTranslationToPortuguese() ? `Portuguese: ${this.getBodySuffix()}` : 'English: ';
  }

  getBodySuffix = () => ` [${this.gender}, ${this.number}]`;

  getDescription = () => {
    if (this.isTranslationToPortugueseFromHearing()) return 'Listen...';
    if (this.isTranslationToPortuguese()) {
      return `English: ${this.other.english}`;
    }
    return `Portuguese: ${this.other.portuguese.singular.masculine}`;
  };

  getTranslation = () => this.other.english;

  getCorrectAnswer = () =>
    this.isTranslationToPortuguese() ? this.other.portuguese[this.number][this.gender] : this.other.english;

  isAnswerCorrect(answer: string): boolean {
    return this.getCorrectAnswer().toLowerCase() === answer.toLowerCase();
  }

  getRetryPrompt = () =>
    this.isTranslationToPortuguese() ? this.getCorrectAnswer() : this.other.portuguese[this.number][this.gender];

  getBaseWord() {
    return this.other;
  }

  getBaseWordType(): BaseWordType {
    return BaseWordType.OTHER;
  }

  getBaseWordAsString(): string {
    return this.getBaseWord().portuguese.singular.masculine;
  }

  equal = (other: OtherGenderTranslationExercise) =>
    other.exerciseType === 'OtherWithGenderTranslation' &&
    this.gender === other.gender &&
    this.number === other.number &&
    this.other.portuguese.singular.masculine === other.other.portuguese.singular.masculine &&
    this.other.portuguese.singular.feminine === other.other.portuguese.singular.feminine &&
    this.other.portuguese.plural.masculine === other.other.portuguese.plural.masculine &&
    this.other.portuguese.plural.feminine === other.other.portuguese.plural.feminine &&
    this.translationType === other.translationType;
}
