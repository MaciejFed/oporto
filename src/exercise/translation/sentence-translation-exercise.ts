import { BaseWordType, ExerciseType } from '../exercise';
import { Comparable } from '../../common/common';
import { Sentence } from '../../repository/exercises-repository';
import { getRandomSentence } from '../../service/translation';
import { TranslationExercise, TranslationType } from './translation-exercise';
import { RatioRange } from '../../service/progress/progress';

export class SentenceTranslationExercise extends TranslationExercise implements Comparable {
  exerciseType: ExerciseType;
  sentence: Sentence;

  constructor() {
    super();
    this.exerciseType = 'SentenceTranslation';
    this.sentence = getRandomSentence();
  }

  static new(sentence: Sentence, translationType: TranslationType): SentenceTranslationExercise {
    const sentenceTranslationExercise = new SentenceTranslationExercise();
    sentenceTranslationExercise.sentence = sentence;
    sentenceTranslationExercise.translationType = translationType;

    return sentenceTranslationExercise;
  }

  isTranslationSubjectEqual(translationExercise: TranslationExercise): boolean {
    if (!(translationExercise instanceof SentenceTranslationExercise)) return false;
    return this.sentence.portuguese === translationExercise.sentence.portuguese;
  }

  getBodyPrefix(): string {
    return this.isTranslationToPortuguese() ? 'Portuguese: ' : 'English: ';
  }

  getBodySuffix = () => '';

  getDescription = () => {
    if (this.isTranslationToPortugueseFromHearing()) return 'Listen...';
    if (this.isTranslationToPortuguese()) {
      return `English: ${this.sentence.english}`;
    }
    return `Portuguese: ${this.sentence.portuguese}`;
  };

  getTranslation = () => (this.isTranslationToPortugueseFromHearing() ? this.sentence.english : undefined);

  getCorrectAnswer = () => (this.isTranslationToPortuguese() ? this.sentence.portuguese : this.sentence.english);

  isAnswerCorrect(answer: string): boolean {
    return this.getCorrectAnswer().toLowerCase().replace('?', '') === answer.toLowerCase().replace('?', '');
  }

  getRetryPrompt = () => (this.isTranslationToPortuguese() ? this.getCorrectAnswer() : this.sentence.portuguese);

  public getMaxProgressRange(): RatioRange {
    return '80-100';
  }

  getBaseWord = () => undefined;

  getBaseWordType(): BaseWordType | undefined {
    return undefined;
  }

  getBaseWordAsString = () => undefined;

  equal = (other: SentenceTranslationExercise) =>
    other.exerciseType === 'SentenceTranslation' &&
    this.sentence.portuguese === other.sentence.portuguese &&
    this.sentence.english === other.sentence.english &&
    this.sentence.sentenceType === other.sentence.sentenceType &&
    this.translationType === other.translationType;
}
