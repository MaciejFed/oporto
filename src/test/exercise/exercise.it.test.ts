/* eslint-disable @typescript-eslint/no-var-requires */
import { simulateContinueButton, simulateTyping } from '../util';
import { withBaseMocks } from '../base-mocks';

const runExerciseSnapshotTest = (exercises: any[]) => {
  const { mockGenerateExercisesForSession, eventProcessor, SessionManager, Output } = withBaseMocks(true);

  const answers = exercises.map((exercise) => exercise.getCorrectAnswer()).reverse();

  mockGenerateExercisesForSession.mockImplementation(() => {
    return exercises;
  });

  const sessionManager = new SessionManager(eventProcessor, 2, false, () => true);
  eventProcessor.emit('APP_STARTED');

  const outputs = answers.map((answer) => {
    simulateTyping(answer.substring(0, answer.length - 2), true);
    simulateTyping(answer, true);
    const output = Output.getOutput();

    simulateContinueButton();

    return output;
  });

  const eventHistory = eventProcessor.eventHistory.map((event: { event: any; args: any }) =>
    `${event.event} ${event.args ? [event.args] : ''}`.trimEnd()
  );
  expect(eventHistory.includes('APP_FINISHED')).toBe(true);
  outputs.forEach((output) => {
    expect(output).toMatchSnapshot();
  });
  expect(eventHistory).toMatchSnapshot();
};

describe('Exercises Integration Snapshots', () => {
  beforeEach(() => {});

  afterEach(() => {
    jest.resetModules();
    process.stdin.removeAllListeners();
  });
  const readAll = require('../../repository/exercises-repository').readAll;

  it('NounTranslation', () => {
    const NounTranslationExercise =
      require('../../exercise/translation/noun-translation-exercise').NounTranslationExercise;
    const nounTranslationExerciseToPortuguese = NounTranslationExercise.new(readAll().nouns[0], 'toPortuguese');
    const nounTranslationExerciseToEnglish = NounTranslationExercise.new(readAll().nouns[0], 'toEnglish');

    runExerciseSnapshotTest([nounTranslationExerciseToPortuguese, nounTranslationExerciseToEnglish]);
  });

  it('SentenceTranslation', () => {
    const SentenceTranslationExercise =
      require('../../exercise/translation/sentence-translation-exercise').SentenceTranslationExercise;
    const sentenceTranslationExerciseToPortuguese = SentenceTranslationExercise.new(
      readAll().sentences[0],
      'toPortuguese'
    );
    const sentenceTranslationExerciseToEnglish = SentenceTranslationExercise.new(readAll().sentences[0], 'toEnglish');

    runExerciseSnapshotTest([sentenceTranslationExerciseToPortuguese, sentenceTranslationExerciseToEnglish]);
  });

  it('VerbExercise', () => {
    const VerbExercise = require('../../exercise/verb-exercise').VerbExercise;
    const verbExercisePresent = VerbExercise.new(readAll().verbs[0], 'Tu', 'presentSimple');
    const verbExercisePast = VerbExercise.new(readAll().verbs[0], 'Tu', 'pastPerfect');

    runExerciseSnapshotTest([verbExercisePresent, verbExercisePast]);
  });
});
