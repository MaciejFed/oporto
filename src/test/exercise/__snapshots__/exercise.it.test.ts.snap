// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exercises Integration Snapshots NounTranslation 1`] = `
Array [
  "APP_STARTED",
  "TERMINAL_CLEARED preExerciseClear",
  "EXERCISE_STARTED",
  "TERMINAL_CLEARED preExerciseClear",
  "EXERCISE_DESCRIPTION_PRINTED Portuguese: o chá",
  "EXERCISE_BODY_PRINTED [object Object]",
  "KEY_PRESSED t",
  "KEY_PRESSED e",
  "KEY_PRESSED a",
  "ANSWER_SUBMITTED voice",
  "ANSWER_CHECKED [object Object]",
  "KEY_PRESSED",
  "EXERCISE_NEXT",
  "EXERCISE_STARTED",
  "TERMINAL_CLEARED preExerciseClear",
  "EXERCISE_DESCRIPTION_PRINTED English: tea",
  "EXERCISE_BODY_PRINTED [object Object]",
  "KEY_PRESSED o",
  "KEY_PRESSED",
  "KEY_PRESSED c",
  "KEY_PRESSED h",
  "KEY_PRESSED á",
  "ANSWER_SUBMITTED voice",
  "ANSWER_CHECKED [object Object]",
  "KEY_PRESSED",
  "EXERCISE_NEXT",
  "APP_FINISHED",
]
`;

exports[`Exercises Integration Snapshots NounTranslation 2`] = `
"IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
                                                                      
                                                                      
                                                                      
                                                                      
 Portuguese: o chá                                                    
 English: tea                                                         
 Correct! [voice]                                                     
                                                                      
                                                                      
 Press key to continue...                                             
 r - repeat the answer                                                
                                                                      
                                                                      
Last Month Results:                                                   
                                                                      
Last Time Attempted: 2022-7-19 12:36 [in 0 seconds]                   
                                                                      
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      "
`;

exports[`Exercises Integration Snapshots NounTranslation 3`] = `
"IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
                                                                      
                                                                      
                                                                      
                                                                      
 English: teao chá                                                    
 Portuguese: o chá                                                    
 Correct! [voice]                                                     
                                                                      
                                                                      
 Press key to continue...                                             
 r - repeat the answer                                                
                                                                      
                                                                      
Last Month Results:                                                   
                                                                      
Last Time Attempted: 2022-7-19 12:36 [in 0 seconds]                   
                                                                      
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      "
`;

exports[`Exercises Integration Snapshots SentenceTranslation 1`] = `
Array [
  "APP_STARTED",
  "TERMINAL_CLEARED preExerciseClear",
  "EXERCISE_STARTED",
  "TERMINAL_CLEARED preExerciseClear",
  "EXERCISE_DESCRIPTION_PRINTED Portuguese: Boa ideia",
  "EXERCISE_BODY_PRINTED [object Object]",
  "KEY_PRESSED G",
  "KEY_PRESSED o",
  "KEY_PRESSED o",
  "KEY_PRESSED d",
  "KEY_PRESSED",
  "KEY_PRESSED i",
  "KEY_PRESSED d",
  "KEY_PRESSED e",
  "KEY_PRESSED a",
  "ANSWER_SUBMITTED voice",
  "ANSWER_CHECKED [object Object]",
  "KEY_PRESSED",
  "EXERCISE_NEXT",
  "EXERCISE_STARTED",
  "TERMINAL_CLEARED preExerciseClear",
  "EXERCISE_DESCRIPTION_PRINTED English: Good idea",
  "EXERCISE_BODY_PRINTED [object Object]",
  "KEY_PRESSED B",
  "KEY_PRESSED o",
  "KEY_PRESSED a",
  "KEY_PRESSED",
  "KEY_PRESSED i",
  "KEY_PRESSED d",
  "KEY_PRESSED e",
  "KEY_PRESSED i",
  "KEY_PRESSED a",
  "ANSWER_SUBMITTED voice",
  "ANSWER_CHECKED [object Object]",
  "KEY_PRESSED",
  "EXERCISE_NEXT",
  "APP_FINISHED",
]
`;

exports[`Exercises Integration Snapshots SentenceTranslation 2`] = `
"IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
                                                                      
                                                                      
                                                                      
                                                                      
 English: teao chá                                                    
 Portuguese: o chá                                                    
 Correct! [voice]                                                     
                                                                      
                                                                      
 Press key to continue...                                             
 r - repeat the answer                                                
                                                                      
                                                                      
Last Month Results:                                                   
                                                                      
Last Time Attempted: 2022-7-19 12:36 [in 0 seconds]                   
                                                                      
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      "
`;

exports[`Exercises Integration Snapshots SentenceTranslation 3`] = `
"IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
                                                                      
                                                                      
                                                                      
                                                                      
 English: teao chá                                                    
 Portuguese: o chá                                                    
 Correct! [voice]                                                     
                                                                      
                                                                      
 Press key to continue...                                             
 r - repeat the answer                                                
                                                                      
                                                                      
Last Month Results:                                                   
                                                                      
Last Time Attempted: 2022-7-19 12:36 [in 0 seconds]                   
                                                                      
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      "
`;

exports[`Exercises Integration Snapshots VerbExercise 1`] = `
Array [
  "APP_STARTED",
  "TERMINAL_CLEARED preExerciseClear",
  "EXERCISE_STARTED",
  "TERMINAL_CLEARED preExerciseClear",
  "EXERCISE_DESCRIPTION_PRINTED Infinitive: to see",
  "EXERCISE_BODY_PRINTED [object Object]",
  "KEY_PRESSED v",
  "KEY_PRESSED i",
  "KEY_PRESSED s",
  "KEY_PRESSED t",
  "KEY_PRESSED e",
  "ANSWER_SUBMITTED voice",
  "ANSWER_CHECKED [object Object]",
  "KEY_PRESSED",
  "EXERCISE_NEXT",
  "EXERCISE_STARTED",
  "TERMINAL_CLEARED preExerciseClear",
  "EXERCISE_DESCRIPTION_PRINTED Infinitive: to see",
  "EXERCISE_BODY_PRINTED [object Object]",
  "KEY_PRESSED v",
  "KEY_PRESSED ê",
  "KEY_PRESSED s",
  "ANSWER_SUBMITTED voice",
  "ANSWER_CHECKED [object Object]",
  "KEY_PRESSED",
  "EXERCISE_NEXT",
  "APP_FINISHED",
]
`;

exports[`Exercises Integration Snapshots VerbExercise 2`] = `
"IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
                                                                      
                                                                      
                                                                      
                                                                      
 English: teao chá                                                    
 Portuguese: o chá                                                    
 Correct! [voice]                                                     
                                                                      
                                                                      
 Press key to continue...                                             
 r - repeat the answer                                                
                                                                      
                                                                      
Last Month Results:                                                   
                                                                      
Last Time Attempted: 2022-7-19 12:36 [in 0 seconds]                   
                                                                      
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      "
`;

exports[`Exercises Integration Snapshots VerbExercise 3`] = `
"IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
                                                                      
                                                                      
                                                                      
                                                                      
 English: teao chá                                                    
 Portuguese: o chá                                                    
 Correct! [voice]                                                     
                                                                      
                                                                      
 Press key to continue...                                             
 r - repeat the answer                                                
                                                                      
                                                                      
Last Month Results:                                                   
                                                                      
Last Time Attempted: 2022-7-19 12:36 [in 0 seconds]                   
                                                                      
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
IMAGE                                                                     
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      
                                                                      "
`;
